presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java

```diff
@@ -357,13 +358,31 @@ public boolean equals(Object o)
 
         TypeSignature other = (TypeSignature) o;
 
+        // TODO remove this hack together with hack from toString()
+        if (magicVarcharEquals(other, this)) {
```

Comment:
Actually this is expected behavior.
If we allow two different signatures to represent same type (which we should eventually not allow) - then we need a hack here.

Originally I hacked parser to parse `"varchar"` as `"varchar(2^32)" but generated errors as plenty of functions operating on varchars are defined to take `"varchar"` as parameter instead `"varchar(x)"`.

When function definitions are fixed we can change this hack - to smaller hack in parser.